src/basic_recipes/contours.jl

```diff
@@ -172,18 +189,63 @@ function plot!(plot::T) where T <: Union{Contour, Contour3d}
 
     replace_automatic!(()-> zrange, plot, :colorrange)
 
-    args = @extract plot (color, colormap, colorrange, alpha)
+    labels, label_attributes, args... = @extract plot (labels, label_attributes, color, colormap, colorrange, alpha)
     level_colors = lift(color_per_level, args..., levels)
-    result = lift(x, y, z, levels, level_colors) do x, y, z, levels, level_colors
+    result = lift(x, y, z, levels, level_colors, labels) do x, y, z, levels, level_colors, labels
         t = eltype(z)
         # Compute contours
         xv, yv = to_vector(x, size(z,1), t), to_vector(y, size(z,2), t)
         contours = Contours.contours(xv, yv, z,  convert(Vector{eltype(z)}, levels))
-        contourlines(T, contours, level_colors)
+        contourlines(T, contours, level_colors, labels)
+    end
+    scene = parent_scene(plot)
+
+    texts = lift(labels, label_attributes, color, result) do labels, label_attributes, color, (_, _, str_pos)
+        labels || return
+        map(str_pos) do (str, (p1, p2, p3))
```

Comment:
Here's the problem in the logic. You're reacting to `labels, label_attributes, color, result` to set the rotation, and you actually do `text!(scene...` within the lift block. You should never do that because you really only want one `text` object, but this is supposed to be updated.

So make an empty `text` plot before this `lift` block which takes an observable of a vector of strings as input. These can be empty at first.

Then you do `onany(labels, label_attributes, whatever_you_need..., scene.camera.projectionview, scene.px_area)` and in that block you first compute the labels and their positions in the scene. Then you update the `text` plot with that. Now, in the same block, you will be able to query the new boundingboxes from the `Vector{GlyphCollection}` underlying the `text` plot because it has just been updated. With these boundingboxes, you null out the affected points in the contour lines copy.

Reply:
Ok, thanks for the hints, I will try moving the `text!` out of the `lift` block and change the approach.

One problem is that you use a single text elements, and I couldn't find a way to access individual text elements i.e. a single glyph collection, to be able to compute individual bboxes.

Reply:
Ok yeah this is not yet so well developed, but I made it work like this:

```julia
s = Scene(camera = campixel!)
points = Point2f[(100, 100), (200, 200)]
t = text!(s, points, text = ["One string", "Another longer string"], rotation = [0, 0.5pi])
glyphcollections = t.plots[1][1][]
Makie.broadcast_foreach(glyphcollections, points, Makie.to_rotation(t.rotation[])) do gc, point, rot
    wireframe!(s, Makie.boundingbox(gc, Makie.to_ndim(Point3f, point, 0), rot))
end
s
```

![image](https://user-images.githubusercontent.com/22495855/207830742-91c2c0e7-583d-48e5-91fb-e99bf94cf71b.png)

Reply:
You might even want to do it differently and compute the boundingbox before rotation, then rotate it, because otherwise the bbox is pretty big for long oblique strings. But that can be an optimization for later.

```julia
s = Scene(camera = campixel!)
points = Point2f[(100, 100), (200, 200)]
t = text!(s, points, text = ["One string", "Another longer string"], rotation = [0, 0.25pi])
glyphcollections = t.plots[1][1][]
Makie.broadcast_foreach(glyphcollections, points, Makie.to_rotation(t.rotation[])) do gc, point, rot
    wireframe!(s, Makie.boundingbox(gc, Makie.to_ndim(Point3f, point, 0), rot))
end
s
```
![image](https://user-images.githubusercontent.com/22495855/207831449-ecfe1b9c-d8a5-4e2f-9c94-440c1dc8f9d9.png)

Reply:
> You might even want to do it differently and compute the boundingbox before rotation, then rotate it, because otherwise the bbox is pretty big for long oblique strings.

Ha yeah, the `30` label in 2d.

Reply:
3D seems super annoying as everything overlaps :)

Reply:
So I've reworked using the approach described in https://github.com/MakieOrg/Makie.jl/pull/2496#discussion_r1049424057.

It still fails 3D:

![contour_labels_2d](https://user-images.githubusercontent.com/13423344/207839706-810c3bfe-4a79-4da4-8f7e-166d3a44ff12.png)
![contour_labels_3d](https://user-images.githubusercontent.com/13423344/207839721-6a118cd0-501e-41a3-8d9f-be27d352c1d8.png)

Reply:
Looks weird, the boundingboxes are half correct, half in the wrong place but in the same configuration, both sets of labels are in the same place but ever so slightly shifted. Didn't you have label placement calculated correctly before?

Reply:
Right, wrong position type.

![contour_labels_3d](https://user-images.githubusercontent.com/13423344/207845812-903e1657-f0f7-4b5e-b05f-3689c776c2f4.png)

Now, were back to the pipeline update problem described in https://github.com/MakieOrg/Makie.jl/pull/2496#issuecomment-1352126622, and the project to `:pixel` space in 3D.

Reply:
You still have duplicated/shifted boundingboxes. How are you drawing those again?

Reply:
`wireframe!(plot, bb, space = :pixel)`, and the bounding box `bb` updated on `lift(scene.camera.projectionview, scene.px_area, labels)`.

Reply:
Also, the `masked_lines = lift(...)` block gets triggered `22` times which is a lot, and might hurt performance on a big dataset.

Reply:
Ooooh ! After reading [this](https://github.com/MakieOrg/Makie.jl/blob/6da4add742f7a3b465b8c5c3272284e8e81e599c/src/makielayout/blocks/label.jl#L20) it seems that dropping the `z` component of the bounding box in `3D` just works !

Makie would benefit from having commented code. Are we supposed to discover things like this on our own ?

![contour_labels_3d](https://user-images.githubusercontent.com/13423344/207918567-6e03c1cd-d5d5-4e59-b5d1-c2adf2236350.png)